### PR TITLE
Do not read position when tar archive stream is unseekable

### LINKED ISSUE #70172: Creating a tar directly into a GZipStream crashes when writing final records
The root cause is that we access the position, which is unsupported by GZipStream (unseekable stream).

https://github.com/dotnet/runtime/blame/218a4b1dad3f73eec8f9898761501c4cfa14698e/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs#L314

The fix would be to detect if the stream is seekable. If it is, then we can set the length to the position. Otherwise, we skip that step.

Thanks @rainersigwald for [reporting this](https://github.com/dotnet/runtime/issues/65951#issuecomment-1145209082).

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes: https://github.com/dotnet/runtime/issues/70172

The position was being read at the very last moment of creating the archive, to set the archive size one byte after the empty records. This causes an exception to be thrown when compressing directly into a gzip stream, which is unseekable (position cannot be accessed).

cc @rainersigwald

### CODE REVIEW

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs [hunk 1]
```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
...
+
+            using TarReader reader = new TarReader(wrapped);
+            TarEntry entry = reader.GetNextEntry();
+            Assert.NotNull(entry);
+            Assert.Equal(TarFormat.Pax, reader.Format);
+            Assert.Equal(TarEntryType.RegularFile, entry.EntryType);
+            Assert.Null(reader.GetNextEntry());
+        }
+
         [Fact]
```
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:104]: We have a test that verifies the behavior of the `TarEntry.DataStream` when reading an unseekable stream:

https://github.com/dotnet/runtime/blob/2389815a1bc810b5b3c39586b570cbaae2f0f3a5/src/libraries/System.Formats.Tar/tests/TarReader/TarReader.GetNextEntry.Tests.cs#L165

But that test wasn't creating an archive using `TarWriter` against an unseekable stream, hence why this bug wasn't caught before. That's why I'm also adding this extra test to write to a generic unseekable stream.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs [hunk 2]
```diff
@@ -311,7 +311,10 @@ private void WriteFinalRecords()
             byte[] emptyRecord = new byte[TarHelpers.RecordSize];
             _archiveStream.Write(emptyRecord);
             _archiveStream.Write(emptyRecord);
-            _archiveStream.SetLength(_archiveStream.Position);
+            if (_archiveStream.CanSeek)
+            {
+                _archiveStream.SetLength(_archiveStream.Position);
```
Comment (stephentoub) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs:316]: What is the purpose of the SetLength call?

File: src/libraries/System.Formats.Tar/tests/CompressedTar.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+using System.IO.Compression;
+using System.IO;
+using Xunit;
+
+namespace System.Formats.Tar.Tests
+{
+    public class CompressedTar_Tests : TarTestsBase
+    {
+        [Fact]
+        public void TarGz_TarWriter_TarReader()
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/CompressedTar.Tests.cs:13]: are these tests related to the fix?

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs [hunk 4]
```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
...
+            MemoryStream stream = new MemoryStream();
+            using WrappedStream wrapped = new WrappedStream(stream, canRead: true, canWrite: true, canSeek: false);
+
+            using (TarWriter writer = new TarWriter(wrapped, leaveOpen: true))
+            {
+                PaxTarEntry paxEntry = new PaxTarEntry(TarEntryType.RegularFile, "file.txt");
+                writer.WriteEntry(paxEntry);
+            } // The final records should get written, and the length should not be set because position cannot be read
+
+            stream.Seek(0, SeekOrigin.Begin); // Rewind the base stream (wrapped cannot be rewinded)
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:94]:
```suggestion
            stream.Seek(0, SeekOrigin.Begin); // Rewind the base stream (wrapped cannot be rewound)
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs [hunk 5]
```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
...
+
+            using TarReader reader = new TarReader(wrapped);
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:96]: I think it's a little clearer to use a using block to match the writer here.

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs [hunk 6]
```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
...
+            TarEntry entry = reader.GetNextEntry();
+            Assert.NotNull(entry);
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:98]: up to you, but I think many would not bother asserting entry is null, if they are going to get a NRE two lines later from `Assert.Equal(TarEntryType.RegularFile, entry.EntryType);.` maybe it's a style thing.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs (hunk 2 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs:316]: I think I can remove it completely.

I was trying to find the place where I initially added: I recall at the very beginning of writing this code, the `FileStream` was flushing the data to disk and writing it with more bytes than expected. Adding `SetLength` helped me fix that. But I don't have an argument right now to defend that behavior, the tests pass after removing the line.

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs (hunk 4 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:94]: TIL 😄

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs [hunk 7]
```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
             Assert.Equal(0, archiveStream.Length);
         }
 
+        [Fact]
+        public void Write_To_UnseekableStream()
+        {
+            MemoryStream stream = new MemoryStream();
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:85]: Nit, when you wrap a stream, might be clearer to name the inner one something like 'inner' not 'stream'.

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs (hunk 6 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:98]: I particularly like that VS Code lets you click on the filepath with the failure and it will take you directly to the assert that is failing. But in the end I think there's not much difference, I still need to look where that null object came from. So I think I can remove it as you suggest.

```
System.Formats.Tar.Tests.TarWriter_Tests.Write_To_UnseekableStream [FAIL]
        Assert.Null() Failure
        Stack Trace:
          /home/carlos/runtime/src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs(101,0): at System.Formats.Tar.Tests.TarWriter_Tests.Write_To_UnseekableStream()
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          /home/carlos/runtime/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs(64,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

vs

```
System.Formats.Tar.Tests.TarWriter_Tests.Write_To_UnseekableStream [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
          /home/carlos/runtime/src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs(102,0): at System.Formats.Tar.Tests.TarWriter_Tests.Write_To_UnseekableStream()
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          /home/carlos/runtime/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs(64,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs (hunk 5 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs:96]: Sure, I can wrap this.

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs (hunk 7 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs]: Fixed.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarWriter.cs]: Cool.

**Review by stephentoub — APPROVED** [review]
